blast/ncbi/doc/man/asn2xml.1
.TH ASN2XML 1 2003-07-12 NCBI "NCBI Tools User's Manual"
.SH NAME
asn2xml \- translate biological data from ASN.1 to XML
.SH SYNOPSIS
.B asn2xml
[\|\fB\-\fP\|]
[\|\fB\-b\ F\fP\|]
[\|\fB\-e\fP\|]
[\|\fB\-i\fP\ \fIfilename\fP\|]
[\|\fB\-l\fP\ \fIfilename\fP\|]
[\|\fB\-o\fP\ \fIfilename\fP\|]
[\|\fB\-s\fP\|]
.SH DESCRIPTION
\fBasn2xml\fP converts an NCBI Seq-entry or Bioseq-set from ASN.1
format to an isomorphic XML representation.  It differs from
\fBasn2asn \-x\fP in that it converts packed sequence data back into
human-readable ASCII text.
.SH OPTIONS
A summary of options is included below.
.TP
\fB\-\fP
Print usage message
.TP
\fB\-b\ F\fP
Input asnfile in text mode
.TP
\fB\-e\fP
Input is a Seq-entry
.TP
\fB\-i\fP\ \fIfilename\fP
Read ASN.1 from \fIfilename\fP (default is stdin)
.TP
\fB\-l\fP\ \fIfilename\fP
Log errors to \fIfilename\fP (default is stderr)
.TP
\fB\-o\fP\ \fIfilename\fP
Write XML to \fIfilename\fP (default is stdout)
.TP
\fB\-s\fP
Input is a Seq-submit
.SH AUTHOR
The National Center for Biotechnology Information.
.SH SEE ALSO
.ad l
.BR asn2asn (1),
.BR asn2ff (1),
.BR asn2gb (1),
.BR asndhuff (1),
README.asn2xml,
ncbixml.txt


QuickTimeStreamingServer/ReleaseNotes.txt

                                 RELEASE NOTES
                             Darwin Streaming Server 

These release notes are for the current development version of Darwin
Streaming Server.

------------------------------------------------------------------------
Date: 8/7/2001
Version: 3.0.1+development [v349]
cvs tag: qtssServer-349

*   Updated build platforms: Solaris 8, Redhat Linux 7.1, FreeBSD 4.3, 
    Mac OS X 10.1
*   Retransmit/overbuffering (Skip Protection) algorithms tuned
    substantially
*   "reliable_udp" attribute defaults to "enabled" 
*   Various WebAdmin fixes for Windows platform

------------------------------------------------------------------------
Date: 8/9/2001
Version: 3.0.1+development [v350]
cvs tag: qtssServer-350

*   Minor bug fixes.

------------------------------------------------------------------------
Date: 8/14/2001
Version: 3.0.1+development [v351]
cvs tag: qtssServer-351

*   QTSSFileModule - enabled caching based on preferences (shared and
    private).
*   Add buffer delay to on demand movies.
*   QTFile - code cleanup.
*   Changed some paths for the installer.
*   Added support for MP3Broadcaster on Windows.

------------------------------------------------------------------------
Date: 8/16/2001
Version: 3.0.1+development [v352]
cvs tag: qtssServer-352

*   Added MP3StreamingModule to all builds.
*   Fixed gdb script to execute QTSS from build dir on MacOSX.

------------------------------------------------------------------------
Date: 8/19/2001
Version: 3.0.1+development [v353]
cvs tag: qtssServer-353

*   Added:
        add_seconds_to_client_buffer_delay -- Increases client buffer
                                delay before playing movie.
        record_movie_file_sdp -- Records a .sdp file for a .mov file.
                                Appears as moviename.mov.sdp
        enable_movie_file_sdp -- Plays a .mov.sdp file for a .mov movie.
*   Fixed server to play .mov.sdp files.
*   QTSSFileModule enables both private and shared movie cache buffers.

------------------------------------------------------------------------
Date: 8/23/2001
Version: 3.0.1+development [v354]
cvs tag: qtssServer-354

*   Added prefs for MP3StreamingModule.
*   Default Reliable UDP dir is / on UNIX and \ on NT
*   Use Enable Reliable UDP pref to disable Reliable UDP
*   Included new Web Admin
*   Changed relay preferences to XML format.
        Old relay config files will not work.

------------------------------------------------------------------------
Date: 08/28/2001
Version: 3.0.1+development [v355]
cvs tag: qtssServer-355

*	Added preference:
	 RTSP_error_message to enable debugging error strings returned in 
	 the content body of RTSP error responses.
	
*	Changed RTSP code to:
	1) require a Cseq field for OPTIONS requests
	2) allow missing Range headers in a PLAY request (defaults to entire movie)
	3) add content length header to Error responses containing error messages
	4) add SDP o= and t= fields to on-demand movie Describe responses

*      Changed Relay to relay incoming ANNOUNCEd broadcasts
       1) All ANNOUNCEd broadcasts, or
       2) All broadcasts from an IP address
       3) A broadcast URL from an IP address

       (a) Can still configure the relay to relay UDP streams
       (b) Can pull an SDP or movie from source and relay it

*      Changed Relay to ANNOUNCE the SDPs to its destinations
       1) for ANNOUNCED sources
       2) for sources that are pulled from the source server

*      Changed Relay config file to be in XML format
       1) relayconfig.xml-Sample will be installed instead of old streamingrelay.conf

*	Added General Settings and Change Password to UI
	1) Require existing password to change to new
	2) Require the user to enter the same password twice
	3) Change the auth cookie so the user won't have to re-authenticate

-------------------------------------------------------------------------------------------------
Date: 8/31/2001
Version: 3.0.1+development [v356]
cvs tag: qtssServer-356

*	 Admin Module change
	 1) Added an admin module pref called "AdministratorGroup" with
	 a default value of "admin"
	 2) Changed Admin Module to allow all users that belong to the
	 AdministratorGroup, instead of allowing only one user called
	 "streamingadmin".
	 3) A default qtgroups gets installed from now. The default qtusers
	 file has user 'streamingadmin' with password 'default'. The default
	 qtgroups file has the group 'admin' with the user 'streamingadmin'.
	 4) The login name for the web admin can be changed by adding the user
	 to the qtusers file and adding the user to the 'admin' group in the
	 qtgroups file.

*	Added "RTSP_debug_printfs" server pref to show the incoming and outgoing RTSP.
	Fixed Session header response to remove ";timeout=" field if the value is not defined.
	Fixed .mov.sdp Describe response to include the Last-Modified and Cache-Control headers.
	Fixed RTSP Session header parser to also stop on a ';'
	Changes shared file buffer size to default to maximum size if the bit-rate is 0 and no fixed 
	buffer size is defined.

-------------------------------------------------------------------------------------------------
Date: 9/04/2001
Version: 3.0.1+development [v357]
cvs tag: qtssServer-357

*	more implementation code for Web admin added.
*	New XML parser for reading prefs.
*	full implementation of MP3 streaming module is included.

-------------------------------------------------------------------------------------------------
Date: 9/06/2001
Version: 3.0.1+development [v359]
cvs tag: qtssServer-359

*	Relay that pulls source from the same server can be setup now
	[rtsp_source SOURCE with own IP address as the source_addr]
   
-------------------------------------------------------------------------------------------------
Date: 9/10/2001
Version: 3.0.1+development [v361]
cvs tag: qtssServer-361

*	UI: play and stop buttons in playlist screen now start/stop playlists immediately

blast/demo/asn2xml.c
/*****************************************************************************
*
* $Id: asn2xml.c,v 1.2 2001/08/10 18:56:42 kans Exp $
*
*   asn2xml.c
*
*      This program is based on asn2asn.c but modified specifically to
*   efficiently convert asn.1 Seq-entry or Bioseq-set into XML. The
*   only real difference with asn2asn -x is that a special step puts
*   the Seq-inst through the object loaders to convert 2 bit and 4 bit
*   encodings to text formats (1 base per letter) that XML fans prefer.
*
*   The defaults are set so that you can pipe the binary asn.1 Bioseq-set
*   in the GenBank release or update.
*
*   cat update.ent | asn2xml | myxmlreader
*
* Author: Jim Ostell <ostell@ncbi.nlm.nih.gov>
*
*****************************************************************************/
#include <seqport.h>
#include <objsub.h>

#define NUMARG 6
Args myargs[NUMARG] = {
	{"Filename for asn.1 input","stdin",NULL,NULL,FALSE,'i',ARG_FILE_IN,0.0,0,NULL},
	{"Input is a Seq-entry","F", NULL ,NULL ,TRUE,'e',ARG_BOOLEAN,0.0,0,NULL},
	{"Input is a Seq-submit","F", NULL ,NULL ,TRUE,'s',ARG_BOOLEAN,0.0,0,NULL},
	{"Input asnfile in binary mode","T",NULL,NULL,TRUE,'b',ARG_BOOLEAN,0.0,0,NULL},
	{"Filename for XML output","stdout", NULL,NULL,TRUE,'o',ARG_FILE_OUT,0.0,0,NULL},
    {"Log errors to file named:",NULL,NULL,NULL,TRUE,'l',ARG_FILE_OUT, 0.0,0,NULL}};

/*****************************************************************************
*
*   Main program loop to read, process, write SeqEntrys
*
*****************************************************************************/
Int2 Main(void)
{
	AsnIoPtr aipout=NULL, aipin;
	AsnTypePtr atp, atp_inst;
	AsnModulePtr amp;
	DataVal dv;
	CharPtr ftype;
	BioseqPtr bsp;

					/* check command line arguments */

	if ( ! GetArgs("asn2xml 1.0",NUMARG, myargs))
		return 1;

					/* load the sequence alphabets  */
					/* (and sequence parse trees)   */
	if (! SeqEntryLoad())
		ErrShow();
	if (! SubmitAsnLoad())
		ErrShow();
	if (! SeqCodeSetLoad())
		ErrShow();
				    /* get pointer to all loaded ASN.1 modules */
	amp = AsnAllModPtr();
	if (amp == NULL)
	{
		ErrShow();
		return 1;
	}

	if (myargs[1].intvalue)
		atp = AsnFind("Seq-entry");
	else if (myargs[2].intvalue)
		atp = AsnFind("Seq-submit");
	else
		atp = AsnFind("Bioseq-set");    /* get the initial type pointers */
	if (atp == NULL)
	{
		ErrShow();
		return 1;
	}

	atp_inst = AsnFind("Bioseq.inst");
	if (atp_inst == NULL)
	{
		ErrShow();
		return 1;
	}

					/* open the ASN.1 input file in the right mode */

	if ((aipin = AsnIoOpen (myargs[0].strvalue, myargs[3].intvalue?"rb":"r"))
          == NULL)
	{
		ErrPostEx(SEV_ERROR,0,0, "Can't open %s", myargs[0].strvalue);
		ErrShow();
		return 1;
	}

					/* open the ASN.1 output file in the right mode */

	if (myargs[4].strvalue != NULL)   /* output desired? */
	{
		ftype = "wx";

		if ((aipout = AsnIoOpen (myargs[4].strvalue, ftype)) == NULL)
		{
			ErrPostEx(SEV_ERROR,0,0, "Can't open %s", myargs[4].strvalue);
			ErrShow();
			return 1;
		}
	}

                                /* log errors instead of die */
	if (myargs[5].strvalue != NULL)
	{
		if (! ErrSetLog (myargs[5].strvalue))
		{
			ErrShow();
			return 1;
		}
		else
			ErrSetOpts (ERR_CONTINUE, ERR_LOG_ON);
	}

	while ((atp = AsnReadId(aipin, amp, atp)) != NULL)
	{
		if (atp == atp_inst)    /* need object loader convert */
		{
			bsp = BioseqNew();  /* need newly initialized bsp */
			BioseqInstAsnRead(bsp, aipin, atp);
			BioseqInstAsnWrite(bsp, aipout, atp);
			bsp = BioseqFree(bsp);
		}
		else
		{
	 		AsnReadVal(aipin, atp, &dv); /* read it */
			AsnWrite(aipout, atp, &dv);	 /* write it */
			AsnKillValue(atp, &dv);      /* free it */
		}
	}

	AsnIoClose(aipin);
	AsnIoClose(aipout);

	return(0);
}


